Mathlib/Lean/Expr/ReplaceRec.lean

```diff
@@ -0,0 +1,57 @@
+/-
+Copyright (c) 2019 Robert Y. Lewis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Simon Hudon, Scott Morrison, Keeley Hoek, Robert Y. Lewis, Floris van Doorn, E.W.Ayers
+-/
+import Lean
+import Mathlib.Util.TermUnsafe
+import Mathlib.Lean.Expr.Traverse
+import Mathlib.Control.Basic
+import Mathlib.Util.MemoFix
+namespace Lean.Expr
+/-!
+# ReplaceRec
+
+We define a more flexible version of `Expr.replace` where we can use recursive calls even when
+replacing a subexpression. We completely mimic the implementation of `Expr.replace`. -/
+
+private def replaceRecCoreM [Monad M] (f? : (Expr → M Expr) → Expr → M (Option Expr))
```

Comment:
inline?

Reply:
Oops, apparently my comment was too cryptic.  My suggestion was to remove the `replaceRecCoreM` definition, and put its body directly in `replaceRecM`.

Reply:
ahhhh, fixed